File: p08_speed/primes.py
# Count the first N primes by simple trial division.

import sys
import time


def is_prime(n):
    if n < 2:
        return False
    d = 2
    while d * d <= n:
        if n % d == 0:
            return False
        d += 1
    return True


def main():
    n = int(sys.argv[1]) if len(sys.argv) > 1 else 100000

    start = time.perf_counter()

    count = 0
    last = 0
    candidate = 1

    while count < n:
        candidate += 1
        if is_prime(candidate):
            count += 1
            last = candidate

    seconds = time.perf_counter() - start

    print(f"Python : prime #{n} = {last} in {seconds:.3f} s")


main()

File: p08_speed/primes.c
// Count the first N primes by simple trial division.

#include <stdio.h>
#include <stdlib.h>
#include <time.h>

int is_prime(int n) {

  if (n < 2) {
    return 0;
  }

  for (int d = 2; d * d <= n; d++) {
    if (n % d == 0) {
      return 0;
    }
  }

  return 1;
}

int main(int argc, char *argv[]) {

  int n = 100000;
  if (argc > 1) {
    n = atoi(argv[1]);
  }

  clock_t start = clock();

  int count = 0;
  int last = 0;
  int candidate = 1;

  while (count < n) {
    candidate++;
    if (is_prime(candidate)) {
      count++;
      last = candidate;
    }
  }

  double seconds = (double)(clock() - start) / CLOCKS_PER_SEC;

  printf("C      : prime #%d = %d in %.3f s\n", n, last, seconds);

  return 0;
}
